src/main/kotlin/de/uniwuerzburg/omod/core/GridFactory.kt

```diff
@@ -174,4 +183,187 @@ fun makeClusterGrid(focusAreaPrecision: Double, buildings: List<Building>,
+}
+
+/**
+ *     // Get cluster in buffer area with gradually declining resolution
+ */
+fun getBufferBuildingClusters (focusAreaPrecision: Double, focusAreaBoundary: Geometry, bufferBuildings: List<Building>,
+                               geometryFactory: GeometryFactory, transformer: CRSTransformer, startID: Int) : MutableList<Cell>  {
+
+    val cells = mutableListOf<Cell>()
+
+    // Determine distance to focus area and level of detail
+    val buildingGroups = mutableMapOf<Int, MutableList<Building>>()
+    for (building in bufferBuildings) {
+        val distance = focusAreaBoundary.distance(building.point).toInt()
+
+        // Quadratic precision decrease by distance of number of clusters in 10km chunks
+        val nClusterDivisor = 2 + (distance / 10_000).toDouble().pow(2).toInt()
+
+        if(buildingGroups.containsKey(nClusterDivisor)) {
+            buildingGroups[nClusterDivisor]!!.add(building)
+        } else {
+            buildingGroups[nClusterDivisor] = mutableListOf(building)
+        }
+    }
+
+    for ((nClusterDivisor, buildingsAtDistance) in buildingGroups) {
+        cells.addAll(
+            cluster(
+                focusAreaPrecision*nClusterDivisor, buildingsAtDistance, geometryFactory, transformer, startID)
+        )
+    }
+
+    return cells
+}
+
+/**
+ * Group buildings into traffic assignment zones (cells) with predefined geometry.
+ * Calculates cell centroids using building centroids
+ * Cells outside Focus area are not considered yet...
+ *
+ * @param tazFile geojson file with geometries of traffic assignment zones (cells)
+ * @param buildings the buildings which are assigned to cells
+ * @param transformer CRS transformer to use
+ *
+ * @return routing cells
+ */
+@OptIn(ExperimentalSerializationApi::class)
+@Suppress("unused")
+fun makeClusterGridFromFile(tazFile: File, buildings: List<Building>, transformer: CRSTransformer ) : List<Cell> {
+
+    // Get TAZs from file
+    val taz : GeoJsonNoProperties = json.decodeFromString(tazFile.readText(Charsets.UTF_8))
+
+    // extract shape of each zone
+    var zones = arrayOf<Geometry>()
+    if (taz is GeoJsonFeatureCollectionNoProperties) {
+        zones = taz.features.map { it.geometry.toJTS(geometryFactory) }.toTypedArray()
+    }
+
+    // Get Buildings in FocusArea
+    val focusAreaBuildings = buildings.filter { it.inFocusArea }
+
+    // Init empty cells
+    val grid = mutableListOf<Cell>()
+
+    // Init empty cell buildings arrays (one for each zone)
+    val cellBuildingLists = MutableList(zones.size) { mutableListOf<Building>() }
+
+    /*
+    val cellBuildingsTree = HPRtree()
+    for (building in focusAreaBuildings) {
+        cellBuildingsTree.insert(building.geometry.envelopeInternal, building)
+    }
+    */
+
+
+    // Structure for aggregated cell information
+    val simpleTAZs = mutableListOf<OutputSimpleTAZ>()
+
+    // Fill Cells with buildings
+
+    logger.info("Fill Focus Area Cells...")
+    for (b in focusAreaBuildings) {
```

Comment:
Slow! Use:

for (odZone in odZones) {
val zoneBuildings = mutableListOf<Building>()

fastCovers(odZone.geometry, listOf(10000.0, 5000.0, 1000.0), geometryFactory,
    ifNot = { },
    ifDoes = { e ->
        zoneBuildings.addAll(kdTree.query(e).map { ((it as KdNode).data as Building) })
    },
    ifUnsure = { e ->
        zoneBuildings.addAll(kdTree.query(e).map { ((it as KdNode).data as Building) }
            .filter { odZone.geometry.contains(it.point) })
    }
)